File: ctci/tmp/chapter_4_trees_graphs/js/4.10-checkSubtree.js
const { TreeNode } = require('./tree')

/**
 * Create a string representation of binary with a pre-order traversal.
 * All null attributes are replaced with 'X'.
 * @param   {TreeNode}  root
 * @returns {string}
 */
const getOrderString = root => {
  if (!root) return 'X'
  return root.value + getOrderString(root.left) + getOrderString(root.right)
}

/**
 * Determines if a tree contains another tree.
 * @param   {TreeNode}  firstRoot
 * @param   {TreeNode}  secondRoot
 * @returns {boolean}
 */
const containsTree = (firstRoot, secondRoot) => {
  const str1 = getOrderString(firstRoot)
  const str2 = getOrderString(secondRoot)
  return str1.includes(str2)
}

const n1 = new TreeNode(1)
const n2 = n1.addLeft(2)
const n3 = n1.addRight(3)
const n4 = n2.addLeft(4)
const n5 = n2.addRight(5)
const n6 = n3.addLeft(6)
const n8 = n4.addLeft(8)
const n9 = n4.addRight(9)
const n10 = n5.addLeft(10)
const n11 = n5.addRight(11)
const n12 = n6.addLeft(12)
const n13 = n6.addRight(13)

const m1 = new TreeNode(2)
const m2 = m1.addLeft(4)
const m3 = m1.addRight(5)
const m4 = m2.addLeft(8)
const m5 = m2.addRight(9)
const m6 = m3.addLeft(10)
const m7 = m3.addRight(11)

console.log(containsTree(n1, m1)) // true
m3.right = null
console.log(containsTree(n1, m1)) // false

File: ctci/tmp/chapter_4_trees_graphs/python/4.10-check_subtree.py
TreeNode = __import__('tree').TreeNode


def get_order_string(node: TreeNode) -> str:
    """
    Create a string representation of binary with a pre-order traversal. All
    null attributes are replaced with 'X'.
    """
    if not node:
        return "X"
    return str(node.value) + get_order_string(node.left) +\
        get_order_string(node.right)


def contains_tree(first_root: TreeNode, second_root: TreeNode) -> bool:
    """
    Determines if a tree contains another tree.
    """
    str1 = get_order_string(first_root)
    str2 = get_order_string(second_root)
    return str2 in str1


if __name__ == "main":
    n1 = TreeNode(10)
    n2 = n1.add_left(5)
    n3 = n1.add_right(-3)
    n4 = n2.add_left(3)
    n5 = n2.add_right(1)
    n7 = n3.add_right(11)
    n8 = n4.add_left(3)
    n9 = n4.add_right(-2)
    n11 = n5.add_right(2)

    m1 = TreeNode(5)
    m2 = m1.add_left(3)
    m3 = m1.add_right(1)
    m4 = m2.add_left(3)
    m5 = m2.add_right(-2)
    m7 = m3.add_right(2)

    print(contains_tree(n1, m1))  # true
    m3.add_right(0)
    print(contains_tree(n1, m1))  # false
